modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/SqlPlanHistoryIntegrationTest.java

```diff
@@ -0,0 +1,718 @@
+
+        List<String> cmds = qrysInfo.get1();
+
+        task.accept(cmds);
+
+        checkSqlPlanHistoryDml(3, qrysInfo.get2());
+    }
+
+    /** Returns current SQL plan history on the query node. */
+    public Map<SqlPlan, Long> getSqlPlanHistory() {
+        return queryNode().context().query().runningQueryManager().planHistoryTracker().sqlPlanHistory();
+    }
+
+    /**
+     * Checks SQL plan history for select queries.
+     *
+     * @param size Number of SQL plan entries expected to be in the history.
+     */
+    public void checkSqlPlanHistory(int size) {
+        if (!isReducePhase && isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME)
+            size = 0;
+
+        executeHistoryCheck(size, plans -> plans);
+    }
+
+    /**
+     * Checks SQL plan history for DML operations.
+     *
+     * @param size Number of SQL plan entries expected to be in the history.
+     * @param isSimpleQry Simple query flag.
+     */
+    public void checkSqlPlanHistoryDml(int size, boolean isSimpleQry) {
+        executeHistoryCheck(size, plans -> {
+            if (sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME) {
+                String check;
+
+                if (isSimpleQry)
+                    check = "no SELECT queries have been executed.";
+                else
+                    check = "the following " + (loc ? "local " : "") + "query has been executed:";
+
+                plans = plans.entrySet().stream()
+                    .filter(e -> e.getKey().plan().contains(check))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+            }
+
+            return plans;
+        });
+    }
+
+    /**
+     * @param size Number of SQL plan entries expected to be in the history.
+     * @param func Function to get the correct collection of plans for further checking.
+     */
+    public void executeHistoryCheck(int size, Function<Map<SqlPlan, Long>, Map<SqlPlan, Long>> func) {
+        Map<SqlPlan, Long> sqlPlans = getSqlPlanHistory();
+
+        assertNotNull(sqlPlans);
+
+        sqlPlans = func.apply(sqlPlans);
+
+        checkMetrics(size, sqlPlans);
+    }
+
+    /**
+     * Checks metrics of provided SQL plan history entries.
+     *
+     * @param size Number of SQL plan entries expected to be in the history.
+     * @param sqlPlans Sql plans recorded in the history.
+     */
+    public void checkMetrics(int size, Map<SqlPlan, Long> sqlPlans) {
+        if (size == 1 && sqlPlans.size() == 2) {
+            List<Map.Entry<SqlPlan, Long>> sortedPlans = new ArrayList<>(sqlPlans.entrySet()).stream()
+                .sorted(Comparator.comparing(Map.Entry::getValue))
+                .collect(Collectors.toList());
+
+            String plan1 = sortedPlans.get(0).getKey().plan();
+            String plan2 = sortedPlans.get(1).getKey().plan();
+
+            assertTrue(plan2.contains(plan1) && plan2.contains("/* scanCount"));
```

Comment:
Why do we have two plans?

Reply:
With local queries, when the same query is executed multiple times, H2 adds the `/* scanCount: ...` suffix to the plan string. As a result, several entries might be added to the view for the same query, differing only by the aforementioned suffix.